archive_query_log/parsers/warc_direct_answers.py

```diff
@@ -0,0 +1,272 @@
+from functools import cache
+from itertools import chain
+from typing import Iterable, Iterator
+from urllib.parse import urljoin
+from uuid import uuid5
+
+from click import echo
+from elasticsearch_dsl import Search
+from elasticsearch_dsl.function import RandomScore
+from elasticsearch_dsl.query import FunctionScore, Term, RankFeature, Exists
+# noinspection PyProtectedMember
+from lxml.etree import _Element, tostring  # nosec: B410
+from tqdm.auto import tqdm
+from warc_s3 import WarcS3Store
+
+from archive_query_log.config import Config
+from archive_query_log.namespaces import NAMESPACE_WARC_DIRECT_ANSWER_PARSER, \
+    NAMESPACE_RESULT
+from archive_query_log.orm import Serp, InnerParser, InnerProviderId, \
+    WarcDirectAnswerParserType, WarcDirectAnswerParser, WarcLocation, DirectAnswer, \
+    Result, InnerSerp, DirectAnswerId, InnerDownloader
+from archive_query_log.parsers.warc import open_warc
+from archive_query_log.parsers.xml import parse_xml_tree, safe_xpath
+from archive_query_log.utils.es import safe_iter_scan, update_action
+from archive_query_log.utils.time import utc_now
+
+
+def add_warc_direct_answer_parser(
+        config: Config,
+        provider_id: str | None,
+        url_pattern_regex: str | None,
+        priority: float | None,
+        parser_type: WarcDirectAnswerParserType,
+        xpath: str | None,
+        big_box_xpath: str | None,
+        small_box_xpath: str | None,
+        right_box_xpath: str | None,
+) -> None:
+    if priority is not None and priority <= 0:
+        raise ValueError("Priority must be strictly positive.")
+    if parser_type == "xpath":
+        if xpath is None:
+            raise ValueError("No XPath given.")
+    else:
+        raise ValueError(f"Invalid parser type: {parser_type}")
+    parser_id_components = (
+        provider_id if provider_id is not None else "",
+        url_pattern_regex if url_pattern_regex is not None else "",
+        str(priority) if priority is not None else "",
+    )
+    parser_id = str(uuid5(
+        NAMESPACE_WARC_DIRECT_ANSWER_PARSER,
+        ":".join(parser_id_components),
+    ))
+    parser = WarcDirectAnswerParser(
```

Comment:
The constructor parameters need to match the fields from the `WarcDirectAnswerParser` class: https://github.com/webis-de/archive-query-log/blob/615d0f41a5ff82fcbb799186167735314e86e9dc/archive_query_log/orm.py#L459